### ISSUE TITLE
Temperature compensation

### ISSUE THREAD

**Issue by CamDavidsonPilon** [Pioreactor/pioreactor#143]
- [ ] map the LED in the config.ini to a `TemperatureCompensator` in `od_reading.py`
 - [ ] adjust the temperature compensation if the user runs the action locally
 - [ ] decide how to deal with other IR LEDs, besides the one currently in use. 
    - does the coefficient come from the cloud?

**Comment by CamDavidsonPilon** [comment]
This can be done globally (i.e. pioreactor software provides values), but can also be done locally, per unit. If the user supplies a static media (something that will not change with time or temperature... tbd), then the Pioreactor unit can gradually increase temperature and record OD, and save those values locally.

**Comment by CamDavidsonPilon** [comment]
If/When users to a local compensation, this is useful data for Pioreactor to receive, to improve a global model. Introduce sharing.

**Comment by CamDavidsonPilon** [comment]
Some possible routines:

1. Turn on OD Reading. Turn on TempControl, and starting measuring temps every 10m (as usual). Slowly increase the duty cycle over time to get a range of temperatures. When a new temp is logged, take average OD Reading. At the end, we would have a map of {temp: od_reading}. This can be normalized using the od reading at (estimated) 30C?


Question: 
1. Interpolation is easy, but we will have problems with extrapolation. 
2. How correct is a linear model between temp and od_reading?

**Comment by CamDavidsonPilon** [comment]
Some data from a 45 deg angle, using water as media
```
{
"28.0625": 0.04119389212794672,
"29.5": 0.04090225034389634,
"30.8125": 0.04023479874928575,
"32.0625": 0.0399462981743179,
"33.84375": 0.03961079326678997,
"35.28125": 0.03911513127520956,
"36.65625": 0.038698652002730226,
"38.46875": 0.03879344055171691,
"39.96875": 0.03836538225908753,
"41.59375": 0.03835397601722154,
"43.5625": 0.03776025182761884,
"44.25": 0.03765829275001045,
"44.9375": 0.03774649439696316
}

{
"29.09375": 0.040530522579926856,
"30.0625": 0.04072297896123767,
"31.4375": 0.04033742349454258,
"32.5": 0.04022613655477703,
"33.84375": 0.03977738506249165,
"35.40625": 0.03965389175764982,
"37.03125": 0.03935981091256296,
"38.78125": 0.03882460608171503,
"40.03125": 0.038477954433653386,
"41.34375": 0.0381281580309234,
"43.21875": 0.03738528435694815,
"44.25": 0.03700028237459407,
"45.0625": 0.03710078979375308
}
```

**Comment by CamDavidsonPilon** [comment]
```
model = Model("od ~ temp + (1|source) + 0", df)
results = model.fit(draws=4000, chains=2, init="adapt_diag"

```

<img width="676" alt="Screen Shot 2021-07-22 at 10 04 24 PM" src="https://user-images.githubusercontent.com/884032/126729624-0251ee2c-7e02-4ef8-895e-d47875ae0d2a.png">

```
                    mean        sd    hdi_3%   hdi_97%  mcse_mean   mcse_sd   ess_bulk    ess_tail     r_hat
temp           -0.000228  0.000008 -0.000242 -0.000211   0.000002  0.000001  25.268351  324.960301  1.082487
1|source_sigma  0.011895  0.000806  0.010533  0.013592   0.000104  0.000079  62.419415   83.026558  1.128541
1|source[0]     0.047497  0.000323  0.046837  0.048029   0.000071  0.000051  25.638553  764.643392  1.077855
1|source[1]     0.047485  0.000325  0.046835  0.048030   0.000072  0.000052  25.386677  498.795760  1.075095
1|source[2]     0.041236  0.000314  0.040617  0.041773   0.000068  0.000049  26.795946  238.924838  1.071815
od_sigma        0.000288  0.000036  0.000229  0.000354   0.000015  0.000011   5.301990   46.222407  1.306146
```

**Comment by CamDavidsonPilon** [comment]
[4trials.csv](https://github.com/Pioreactor/pioreactor/files/6869265/4trials.csv)

```
model = Model("log_od ~ temp + (1|source) + 1", df)
```
```
                    mean        sd    hdi_3%   hdi_97%  mcse_mean   mcse_sd     ess_bulk     ess_tail     r_hat
Intercept      -3.095768  0.072997 -3.244607 -2.956281   0.002095  0.001482  1244.973771  1642.084603  1.000763
temp           -0.006380  0.000218 -0.006786 -0.005954   0.000004  0.000003  3656.849208  3121.194924  1.001106
1|source_sigma  0.143413  0.066029  0.051725  0.268497   0.001790  0.001281  1417.849654  1821.041181  1.000614
1|source[0]     0.088855  0.072714 -0.041376  0.246096   0.002095  0.001482  1230.038712  1635.812337  1.001559
1|source[1]     0.088599  0.072661 -0.041612  0.246671   0.002097  0.001483  1225.475762  1664.932659  1.001707
1|source[2]    -0.085214  0.072645 -0.214420  0.074809   0.002096  0.001483  1225.899973  1654.467393  1.001444
1|source[3]    -0.082030  0.072663 -0.211901  0.077328   0.002094  0.001481  1230.227416  1682.288652  1.001636
log_od_sigma    0.008340  0.000859  0.006754  0.009886   0.000016  0.000011  2953.323681  2446.749514  1.002163

```

<img width="585" alt="Screen Shot 2021-07-23 at 10 03 22 AM" src="https://user-images.githubusercontent.com/884032/126793401-5126cbff-7b66-4308-8c73-191f067c7e52.png">


The important parameter is `temp`, -0.006380. This relates the reduction in OD to temperature, via an exponential. Ex:

If OD at 25 is 2.0, then OD at 30 is 2.0 * exp(-0.006380 * 5) = 1.9372

**Comment by CamDavidsonPilon** [comment]
This entire routine is obsolete now that we are measuring the LED output
